configuration_broker/publisher.cc

```diff
@@ -0,0 +1,100 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+// Contributed by Configured Things Ltd
+
+#include <compartment.h>
+#include <debug.hh>
+#include <fail-simulator-on-error.h>
+#include <thread.h>
+#include <tick_macros.h>
+
+// Expose debugging features unconditionally for this compartment.
+using Debug = ConditionalDebug<true, "Publisher">;
+
+static inline void sleep(const uint32_t mS)
+{
+	Timeout t1{MS_TO_TICKS(mS)};
+	thread_sleep(&t1, ThreadSleepNoEarlyWake);
+}
+
+#include "data.h"
+
+// Compartment can set config values config1 and config2
+#include "config_broker.h"
+#define CONFIG1 "config1"
+DEFINE_WRITE_CONFIG_CAPABILITY(CONFIG1, sizeof(Data))
+#define CONFIG2 "config2"
+DEFINE_WRITE_CONFIG_CAPABILITY(CONFIG2, sizeof(Data))
+
+// Helper to set some dummy config
+void gen_config(SObj        sealedCap,
+                const char *itemName,
+                int         count,
+                const char *token)
+{
+	Debug::log("thread {} Set {}", thread_id_get(), itemName);
+
+	Data *data  = static_cast<Data *>(malloc(sizeof(Data)));
```

Comment:
I don't follow why its redundant - the model now is that the data is allocated on the Broker's heap, with total allocation protected by the size defined in the sealed capability     Wherever the data is held in the publisher until it gets stored by the broker is from this perspective just transient  (it could be some network buffer that the published needs to reuse), and the broker does the copy into its allocated heap via the sandbox

When we were keeping the same heap object and passing the allocation through the various quotas by claims I could see why the callback might be needed - although I'm still unsure about what the publisher might need to keep track between its call to the broker and the callback.

Not having the extra complexity of a callback seemed like a with with this approach   ?

Reply:
With a callback model, the broker allocates and then passes a read-write local capability to the new allocation back to the publisher. The publisher then fills in the contents. When the callback returns, the broker has the only writeable pointer, so the publisher doesn’t need to allocate a buffer that’s copied from but we get the same immutability guarantees (only the broker can mutate the object).

Reply:
So the publisher also need to pass some kind of context object to the broker for whatever data it (the publisher) needs to have access to in the callback ?   I'm kind of missing how otherwise the publisher callback would know what it was that it was going to copy from.

Reply:
In the callback, the publisher has access to its own globals.  You can also add a context object if you want, but it's only really necessary for things that are called simultaneously from multiple threads in the same compartment.